src/go/k8s/pkg/resources/statefulset_update.go

```diff
@@ -0,0 +1,310 @@
+		return err
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) updateUpgradingStatus(
+	ctx context.Context, upgrading bool,
+) error {
+	if !reflect.DeepEqual(upgrading, r.pandaCluster.Status.Upgrading) {
+		r.pandaCluster.Status.Upgrading = upgrading
+		if err := r.Status().Update(ctx, r.pandaCluster); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *StatefulSetResource) partitionUpdateImage(
+	ctx context.Context, sts *appsv1.StatefulSet, newImage string,
+) error {
+	replicas := *sts.Spec.Replicas
+
+	// When a StatefulSet's partition number is set to `i`, only Pods with ordinal
+	// greater than or equal to `i` will be updated.
+	// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#partitions
+	for ordinal := replicas - 1; ordinal >= 0; ordinal-- {
+		// Update() on statefulset has not been called yet in this run, however,
+		// this could be a retry call in which case we skip the current partition.
+		poderr := r.podImageIdenticalToClusterImage(ctx, sts, newImage, ordinal)
+		if poderr == nil {
+			r.logger.Info("Pod already updated, skip", "ordinal", ordinal)
+			continue
+		}
+
+		// Continue only if error is due to Pod not ready, or unchanged image
+		// as an attempt to fix the Pod.
+		if !errors.Is(poderr, errContainerHasWrongImage) && !errors.Is(poderr, errPodNotReady) {
+			return poderr
+		}
+
+		// Before continuing to update the ith Pod, verify that the previously updated
+		// Pod (if any) has rejoined its groups after restarting, i.e., is ready for I/O.
+		if err := r.ensureRedpandaGroupsReady(sts, replicas, ordinal+1); err != nil {
+			return &NeedToReconcileError{RequeueAfter: requeueDuration,
+				Msg:	fmt.Sprintf("redpanda on pod (ordinal: %d) not ready", ordinal)}
+		}
+
+		if err := r.rollingUpdatePartition(ctx, sts, ordinal); err != nil {
+			return err
+		}
+
+		// Restarting the Pod takes enough time to warrant a requeue.
+		return &NeedToReconcileError{RequeueAfter: requeueDuration,
+			Msg:	fmt.Sprintf("wait for pod (ordinal: %d) to restart", ordinal)}
+	}
+
+	// Ensure 0th Pod is ready for I/O before completing the upgrade.
+	if err := r.ensureRedpandaGroupsReady(sts, replicas, 0); err != nil {
+		return &NeedToReconcileError{RequeueAfter: requeueDuration,
+			Msg:	fmt.Sprintf("redpanda on pod (ordinal: %d) not ready", 0)}
+	}
+
+	return nil
+}
+
+// Ensures the Redpanda pod has rejoined its groups after restarting,
+// i.e., is ready for I/O.
+func (r *StatefulSetResource) ensureRedpandaGroupsReady(
+	sts *appsv1.StatefulSet, replicas, ordinal int32,
+) error {
+	if replicas == 0 || ordinal == replicas {
+		return nil
+	}
+
+	headlessServiceWithPort := fmt.Sprintf("%s:%d", r.svc.HeadlessServiceFQDN(),
+		r.pandaCluster.Spec.Configuration.KafkaAPI.Port)
+
+	addresses := []string{fmt.Sprintf("%s-%d.%s", sts.Name, ordinal, headlessServiceWithPort)}
+
+	return queryRedpandaForTopicMembers(addresses, r.logger)
+}
+
+// To be implemented: currently only connects to Kafka API.
```

Comment:
nit: remove this

Reply:
maybe add some context to this? I guess the assumption is that if the consumer can connect, the groups are ready?

Reply:
Unfortunately it's more complex to verify that all groups area read. Here we're saying "at least we can connect and retrieve topics information" - the retrieval happens when `NewConsumer` is called. I will be adding the required logic in this function, or might introduce a /health endpoint in redpanda.

Reply:
I'd have to check the logic, but I suspect sarama will try to connect to all brokers that have joined the cluster, so it might not do quite what you expect, but you probably get away with it when things are going well, at least.

Reply:
@BenPope I'm not providing those brokers to sarama though.. I'm providing the one to test.

@alenkacz: I updated the comment.

Reply:
Note that I'll be revisiting this function and "health" overall.

Reply:
It doesn't care - they're seeds, it will query for all the brokers and then connect to them (maybe). Same reason you can't use rpk to do this - it uses sarama, anyway.

Reply:
btw. this is also problematic when we introduce TLS because then controller will have to have certs to do this call right?